plugins/cache-caffeine/src/test/java/CaffeineHeapCacheTests.java

```diff
@@ -0,0 +1,451 @@
+        caffeineTest.close();
+    }
+
+    public void testInvalidateConcurrently() throws Exception {
+        ToLongBiFunction<ICacheKey<String>, String> weigher = getMockWeigher();
+        MockRemovalListener<String, String> removalListener = new MockRemovalListener<>();
+        ICache<String, String> caffeineTest = new CaffeineHeapCache.Builder<String, String>()
+            .setDimensionNames(List.of(dimensionName))
+            .setExecutor(Runnable::run)
+            .setExpireAfterAccess(TimeValue.MAX_VALUE)
+            .setMaximumWeightInBytes(CACHE_SIZE_IN_BYTES)
+            .setWeigher(weigher)
+            .setRemovalListener(removalListener)
+            .build();
+        int randomKeys = randomIntBetween(10, 100);
+        Map<ICacheKey<String>, String> keyValueMap = new HashMap<>();
+        for (int i = 0; i < randomKeys; i++) {
+            keyValueMap.put(getICacheKey(UUID.randomUUID().toString()), UUID.randomUUID().toString());
+        }
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            caffeineTest.put(entry.getKey(), entry.getValue());
+        }
+        assertEquals(keyValueMap.size(), caffeineTest.count());
+        List<ICacheKey<String>> removedKeyList = new ArrayList<>();
+        Thread[] threads = new Thread[randomKeys];
+        Phaser phaser = new Phaser(randomKeys + 1);
+        CountDownLatch countDownLatch = new CountDownLatch(randomKeys);
+        int j = 0;
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            threads[j] = new Thread(() -> {
+                phaser.arriveAndAwaitAdvance();
+                if (randomBoolean()) {
+                    removedKeyList.add(entry.getKey());
+                    caffeineTest.invalidate(entry.getKey());
+                }
+                countDownLatch.countDown();
+            });
+            threads[j].start();
+            j++;
+        }
+        phaser.arriveAndAwaitAdvance(); // Will trigger parallel puts above.
+        countDownLatch.await();
+
+        for (ICacheKey<String> removedKey : removedKeyList) {
+            assertNull(caffeineTest.get(removedKey));
+        }
+        assertEquals(keyValueMap.size() - removedKeyList.size(), caffeineTest.count());
+        caffeineTest.close();
+    }
+
+    public void testEvictions() throws Exception {
+        ToLongBiFunction<ICacheKey<String>, String> weigher = getMockWeigher();
+        MockRemovalListener<String, String> removalListener = new MockRemovalListener<>();
+        ICache<String, String> caffeineTest = new CaffeineHeapCache.Builder<String, String>()
+            .setDimensionNames(getMockDimensions())
+            .setExecutor(Runnable::run)
+            .setExpireAfterAccess(TimeValue.MAX_VALUE)
+            .setMaximumWeightInBytes(100)
+            .setWeigher(weigher)
+            .setRemovalListener(removalListener)
+            .build();
+        int randomKeys = randomIntBetween(11, 100);
```

Comment:
Should prob set the lower bound to maxWeight / keySize + 1 rather than just 11 so it's clearer where that came from